tools/clang/lib/SPIRV/DebugTypeVisitor.cpp

```diff
@@ -50,150 +53,206 @@ DebugTypeVisitor::lowerCbufferDebugType(const StructType *type,
-  if (instr == nullptr)
-    return false;
-
-  auto *tempType = instr->getTypeTemplate();
-  // It is not a composite type for a resource. It is not an error.
-  if (tempType == nullptr)
-    return true;
-
-  auto &tempParams = tempType->getParams();
-  for (auto &t : tempParams) {
-    auto *loweredParam = lowerToDebugType(t->getSpirvType());
-    if (loweredParam == nullptr)
-      return false;
-    t->setActualType(dyn_cast<SpirvDebugType>(loweredParam));
-    if (!t->getValue()) {
-      auto *debugNone = getDebugInfoNone();
-      if (debugNone == nullptr)
-        return false;
-      t->setValue(debugNone);
+    if (sizeInBits == kUnknownBitSize) {
+      compositeSizeInBits = kUnknownBitSize;
+    } else {
+      compositeSizeInBits = offsetInBits + sizeInBits;
     }
-    setDefaultDebugInfo(t);
   }
-  setDefaultDebugInfo(tempType);
-
-  return true;
+  debugTypeComposite->setMembers(members);
+  debugTypeComposite->setSizeInBits(compositeSizeInBits);
 }
 
-bool DebugTypeVisitor::lowerDebugTypeFunctionForMemberFunction(
-    SpirvDebugInstruction *instr) {
-  auto *fn = dyn_cast<SpirvDebugFunction>(instr);
-  if (fn == nullptr) {
-    emitError("Debug instruction %0 is not a DebugFunction")
-        << instr->getDebugName();
-    return false;
+void DebugTypeVisitor::lowerDebugTypeMembers(
+    SpirvDebugTypeComposite *debugTypeComposite, const StructType *type,
+    const DeclContext *decl) {
+  if (const auto *recordDecl = dyn_cast<RecordDecl>(decl)) {
+    auto fieldIter = recordDecl->field_begin();
+    auto fieldEnd = recordDecl->field_end();
+    unsigned numBases = 0;
+    if (const auto *cxxRecordDecl = dyn_cast<CXXRecordDecl>(recordDecl))
+      numBases = cxxRecordDecl->getNumBases();
+    addDebugTypeForMemberVariables(
+        debugTypeComposite, type,
+        [&fieldIter, &fieldEnd]() {
+          assert(fieldIter != fieldEnd);
+          (void)fieldEnd;
+          auto location = fieldIter->getLocation();
+          ++fieldIter;
+          return location;
+        },
+        numBases);
+  } else if (const auto *hlslBufferDecl = dyn_cast<HLSLBufferDecl>(decl)) {
+    auto subDeclIter = hlslBufferDecl->decls_begin();
+    auto subDeclEnd = hlslBufferDecl->decls_end();
+    addDebugTypeForMemberVariables(
+        debugTypeComposite, type,
+        [&subDeclIter, &subDeclEnd]() {
+          assert(subDeclIter != subDeclEnd);
+          (void)subDeclEnd;
+          auto location = subDeclIter->getLocation();
+          ++subDeclIter;
+          return location;
+        },
+        0);
+  } else {
+    assert(false && "Uknown DeclContext for DebugTypeMember generation");
   }
 
-  if (const auto *fnType = fn->getFunctionType()) {
-    fn->setDebugType(lowerToDebugType(fnType));
-    setDefaultDebugInfo(fn);
-  }
-  if (!fn->getSpirvFunction()) {
-    auto *debugNone = getDebugInfoNone();
-    fn->setDebugInfoNone(debugNone);
+  // Push member functions to DebugTypeComposite Members operand.
```

Comment:
Nit: I think the following loop would have better readability if it was written as a for loop. e.g.

```cpp
for ( auto *subDecl : decl->decls() )
  if (const auto *methodDecl = dyn_cast<FunctionDecl>(subDecl))
    if (auto *dbgFunction = spvContext.getDebugFunctionForDecl(methodDecl))
```

Reply:
Nice. Done.